### DISCUSSION TITLE
Have you heard of "Sacred" for logging experimental parameters?

### DISCUSSION THREAD

**Discussion by beniroquai** [ImSwitch/ImSwitch#165] [Ideas]
Here is the link: https://github.com/IDSIA/sacred
Would this be useful for imswitch too?

**Comment by jacopoabramo** [comment]
This would be an interesting way to substitute shared attributes in order to provide a more solid metadata organization. The problem is that from the documentation it doesn't seem that we can do that without considering the experiment as a "I run a function an gather the data from my experiment". We could get in touch with them asking if there is a nice way to embed this information logging system in a more structured project.

**Reply by jacopoabramo** [reply]
But also... @kasasxav weren't you thinking of an "experiment" interface for ImSwitch to collect a standardized interface depending on microscope setups?

**Reply by xcasas** [reply]
I think this thing of getting all metadata is already done, whenever you save a new file there is a class called sharedAttributes that gathers the metadata from all classes.

https://github.com/ImSwitch/ImSwitch/blob/b05a431320671c2af16b6db35bed6818d4826d70/imswitch/imcommon/model/SharedAttributes.py#L6

**Reply by xcasas** [reply]
update: maybe not the same because it's not connected to doing an experiment.

Yes I have a bit some classes drawn of how I wanted it to be. But it's quite an investment of time. If you are interested I can show it to you. It's a way to generate experiments by linking properties of each experiment with dynamic parameters that determine what the user wants to do with the stages and so on.

**Reply by jacopoabramo** [reply]
> I think this thing of getting all metadata is already done, whenever you save a new file there is a class called sharedAttributes that gathers the metadata from all classes.
> 
> https://github.com/ImSwitch/ImSwitch/blob/b05a431320671c2af16b6db35bed6818d4826d70/imswitch/imcommon/model/SharedAttributes.py#L6

I remember this class, but I honestly never managed to fully understand how to extend this in a manner that one could add its own experiment metadata/configuration info. I'll investigate a bit further how it works then.

> update: maybe not the same because it's not connected to doing an experiment.
> 
> Yes I have a bit some classes drawn of how I wanted it to be. But it's quite an investment of time. If you are interested I can show it to you. It's a way to generate experiments by linking properties of each experiment with dynamic parameters that determine what the user wants to do with the stages and so on.

It would be great if you have some notes about it, so at least I could start thinking about it. Unfortunately there's plenty of stuff for which I need some knowledge transfer :p

**Reply by xcasas** [reply]
So they are all in communication channel so to add them you have to do something like this: https://github.com/ImSwitch/ImSwitch/blob/b05a431320671c2af16b6db35bed6818d4826d70/imswitch/imcontrol/controller/controllers/LaserController.py#L239

**Reply by xcasas** [reply]
Regarding the experiment project I did it together with @nornil , maybe you can team up on that? I can share my notes but I think you could meet up with him since the idea came from both of us based on some of his previous experience with software for microscope control and my experience with ImSwitch. I think he would be happy that someone works on that.

**Reply by jacopoabramo** [reply]
> So they are all in communication channel so to add them you have to do something like this:
> 
> https://github.com/ImSwitch/ImSwitch/blob/b05a431320671c2af16b6db35bed6818d4826d70/imswitch/imcontrol/controller/controllers/LaserController.py#L239

The question is if it's possible to extend access of these shared attributes at a manager level, because I think that this would be helpful for people interested in adding other metadata from the device/controller they implemented. I myself had to add some of these attributes hardcoding them because they needed to be accessed from the device manager. Maybe it's not the best solution thou...?

**Reply by xcasas** [reply]
Yes I mean if you think of a different way that is better you can try and ask the others. I don't think from now on I will be very active in developing changes and new functionality, but I can explain to you how it is done now so that you can compare and take a decision :)

**Reply by jacopoabramo** [reply]
> Regarding the experiment project I did it together with @nornil , maybe you can team up on that? I can share my notes but I think you could meet up with him since the idea came from both of us based on some of his previous experience with software for microscope control and my experience with ImSwitch. I think he would be happy that someone works on that.

I would definetely be up for chatting with him, and your notes are more than welcome!
